src/Cli/dotnet/Commands/Run/FileBasedAppSourceEditor.cs

```diff
@@ -0,0 +1,223 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Immutable;
+using System.Diagnostics;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.DotNet.Cli.Commands.Run;
+
+/// <summary>
+/// A helper to perform edits of file-based app C# source files (e.g., updating the directives).
+/// </summary>
+internal sealed class FileBasedAppSourceEditor
+{
+    public SourceFile SourceFile
+    {
+        get;
+        private set
+        {
+            field = value;
+
+            // Make sure directives are reloaded next time they are accessed.
+            Directives = default;
+        }
+    }
+
+    public ImmutableArray<CSharpDirective> Directives
+    {
+        get
+        {
+            if (field.IsDefault)
+            {
+                field = LoadDirectives(SourceFile);
+                Debug.Assert(!field.IsDefault);
+            }
+
+            return field;
+        }
+        private set
+        {
+            field = value;
+        }
+    }
+
+    public required string NewLine { get; init; }
+
+    private FileBasedAppSourceEditor() { }
+
+    public static FileBasedAppSourceEditor Load(SourceFile sourceFile)
+    {
+        return new FileBasedAppSourceEditor
+        {
+            SourceFile = sourceFile,
+            Directives = LoadDirectives(sourceFile),
+            NewLine = GetNewLine(sourceFile.Text),
+        };
+
+        static string GetNewLine(SourceText text)
+        {
+            // Try to detect existing line endings.
+            string firstLine = text.Lines is [{ } line, ..]
+                ? text.ToString(line.SpanIncludingLineBreak)
+                : string.Empty;
+            return firstLine switch
+            {
+                [.., '\r', '\n'] => "\r\n",
+                [.., '\n'] => "\n",
+                [.., '\r'] => "\r",
+                [.., '\u0085'] => "\u0085",
+                [.., '\u2028'] => "\u2028",
+                [.., '\u2029'] => "\u2029",
+                _ => Environment.NewLine,
+            };
+        }
+    }
+
+    private static ImmutableArray<CSharpDirective> LoadDirectives(SourceFile sourceFile)
+    {
+        return VirtualProjectBuildingCommand.FindDirectives(sourceFile, reportAllErrors: false, DiagnosticBag.Ignore());
+    }
+
+    public void Add(CSharpDirective directive)
+    {
+        TextSpan span = DetermineWhereToAdd(directive, out var append);
```

Comment:
Roslyn has a type `TextChange`, which consists of a `TextSpan` and `string NewText`, and `SourceText.WithChanges()` can take them to created edited versions of source files. Maybe it would be a bit cleaner to use that here?